### ISSUE TITLE
[IMPROVEMENT] Bump CSI sidecar components' version

### ISSUE THREAD

**Issue by PhanLe1010** [longhorn/longhorn#5672]
## Is your improvement request related to a feature? Please describe (👍 if you like this request)

We are currently using csi-provisioner v2.1.2 which has this upstream race bug that sometimes crashes the csi-provisioner pod [https://github.com/kubernetes-csi/external-provisioner/issues/595](https://github.com/kubernetes-csi/external-provisioner/issues/595).

 Considering bump the csi-provisioner version to version 3.x.x to avoid the upstream bug. While we are here, also consider bump other csi-xxx components' version

**Comment by PhanLe1010** [comment]
This improvement should go to v1.4.x and v1.5.x only because Longhorn 1.2.x and v1.3.x have to support Kubernetes 1.18 which [csi-provisioner 3.x.x doesn't support ](https://kubernetes-csi.github.io/docs/external-provisioner.html)

**Comment by ejweber** [comment]
Bumping all of the sidecars is straightforward enough and does not provide any complications in Longhorn v1.5.

Because we announced the deprecation of the v1beta1 VolumeSnapshot API in Longhorn v1.4 (for removal in a future version), we should not upgrade to versions of external-provisioner and external-snapshotter that only support the v1 VolumeSnapshot API.

I created separate PRs for the 1.4.x branches to avoid upgrading external-provisioner and external-snapshotter too far). I created separate PRs for the 1.3.x branches because I slight differences in the versions already in certain files were going to cause merge conflicts.

In general, I think it's valuable to keep the 1.3.x and 1.4.x branches updated with current sidecar images, but I'm not overly opposed to dropping these PRs if we only want to move forward in v1.5.

**Comment by innobead** [comment]
> Bumping all of the sidecars is straightforward enough and does not provide any complications in Longhorn v1.5.
> 
> Because we announced the deprecation of the v1beta1 VolumeSnapshot API in Longhorn v1.4 (for removal in a future version), we should not upgrade to versions of external-provisioner and external-snapshotter that only support the v1 VolumeSnapshot API.

We probably should upgrade this to 1.5 or any concerns?

> I created separate PRs for the 1.4.x branches to avoid upgrading external-provisioner and external-snapshotter too far). I created separate PRs for the 1.3.x branches because I slight differences in the versions already in certain files were going to cause merge conflicts.
> 
> In general, I think it's valuable to keep the 1.3.x and 1.4.x branches updated with current sidecar images, but I'm not overly opposed to dropping these PRs if we only want to move forward in v1.5.

Have you done any testing via running CI (check with @PhanLe1010 ). Technically, it will be a concern to upgrade non-patch versions of any components in our patch releases, so if there are no strong reasons like some mandatory critical issues fixed,  I would suggest only upgrading to 1.5. @PhanLe1010 WDYT? any concerns?

cc @shuo-wu @derekbit

**Comment by innobead** [comment]
@ejweber We Need to ensure all PRs have been verified with local testing and CI before marking those ready to review. About CI, go consult with @PhanLe1010 .

**Comment by ejweber** [comment]
> We probably should upgrade this to 1.5 or any concerns?

I probably should have been more clear. I DID include the potentially breaking upgrades in my pull request to the master branch (intended for 1.5). I did NOT include them in PRs to the 1.4x and 1.3x branches.

**Comment by ejweber** [comment]
> We Need to ensure all PRs have been verified with local testing and CI before marking those ready to review.

I tested the master longhorn/longhorn PR using the following process:

Install Longhorn using the modified Helm chart from the PR branch and command line value overrides (since the new images haven't yet been mirrored to the longhornio Docker repository).

```
helm install -n longhorn-system longhorn \
--set defaultSettings.deletingConfirmationFlag=true \
--set image.csi.attacher.repository=registry.k8s.io/sig-storage/csi-attacher \
--set image.csi.provisioner.repository=registry.k8s.io/sig-storage/csi-provisioner \
--set image.csi.nodeDriverRegistrar.repository=registry.k8s.io/sig-storage/csi-node-driver-registrar \
--set image.csi.resizer.repository=registry.k8s.io/sig-storage/csi-resizer \
--set image.csi.snapshotter.repository=registry.k8s.io/sig-storage/csi-snapshotter \
--set image.csi.livenessProbe.repository=registry.k8s.io/sig-storage/livenessprobe \
./chart
```

Follow the steps in longhorn/longhorn-tests/manager/integration/README.md to deploy the integration tests. Modify longhorn/longhorn-tests/manager/integration/deploy/test.yaml to use the following args (to highlight CSI tests): 

`"-m", "csi", "--skip-recurring-job-test"`

Ensure all tests pass.

```
> kubectl logs -f longhorn-test -c longhorn-test
============================= test session starts ==============================
platform linux -- Python 3.9.15, pytest-5.3.1, py-1.11.0, pluggy-0.13.1 -- /usr/bin/python3.9
cachedir: .pytest_cache
rootdir: /integration, inifile: pytest.ini
plugins: repeat-0.9.1, order-1.0.1
collecting ... collected 384 items / 360 deselected / 24 selected

test_backing_image.py::test_csi_mount_with_backing_image PASSED          [  4%]
test_backing_image.py::test_csi_io_with_backing_image PASSED             [  8%]
test_backing_image.py::test_csi_backup_with_backing_image[s3] PASSED     [ 12%]
test_backing_image.py::test_csi_backup_with_backing_image[nfs] PASSED    [ 16%]
test_csi.py::test_csi_mount PASSED                                       [ 20%]
test_csi.py::test_csi_io PASSED                                          [ 25%]
test_csi.py::test_csi_backup[s3] PASSED                                  [ 29%]
test_csi.py::test_csi_backup[nfs] PASSED                                 [ 33%]
test_csi.py::test_csi_block_volume PASSED                                [ 37%]
test_csi.py::test_csi_offline_expansion PASSED                           [ 41%]
test_csi.py::test_csi_block_volume_online_expansion SKIPPED              [ 45%]
test_csi.py::test_csi_mount_volume_online_expansion SKIPPED              [ 50%]
test_csi.py::test_csi_minimal_volume_size PASSED                         [ 54%]
test_kubernetes.py::test_kubernetes_status PASSED                        [ 58%]
test_kubernetes.py::test_pv_creation PASSED                              [ 62%]
test_kubernetes.py::test_pvc_creation_with_default_sc_set PASSED         [ 66%]
test_kubernetes.py::test_backup_kubernetes_status[s3] PASSED             [ 70%]
test_kubernetes.py::test_backup_kubernetes_status[nfs] PASSED            [ 75%]
test_kubernetes.py::test_delete_with_static_pv PASSED                    [ 79%]
test_kubernetes.py::test_delete_with_provisioned_pv PASSED               [ 83%]
test_kubernetes.py::test_delete_provisioned_pvc PASSED                   [ 87%]
test_recurring_job.py::test_recurring_job_kubernetes_status[s3] SKIPPED  [ 91%]
test_recurring_job.py::test_recurring_job_kubernetes_status[nfs] SKIPPED [ 95%]
test_statefulset.py::test_statefulset_pod_deletion PASSED                [100%]

…

=== 20 passed, 4 skipped, 360 deselected, 53 warnings in 2795.63s (0:46:35) ====

```

I didn't go through these steps with the 1.4.x PRs and 1.4.x longhorn/longhorn tests, but I should have. I'll run them this morning and verify the results.

**Comment by ejweber** [comment]
> Technically, it will be a concern to upgrade non-patch versions of any components in our patch releases, so if there are no strong reasons like some mandatory critical issues fixed, I would suggest only upgrading to 1.5. @PhanLe1010 WDYT? any concerns?

This was something I was unsure about. Unfortunately, the fix we really wanted in this issue (to exernal-provisioner) cannot be backported to previous Longhorn releases. There are small bugfixes in many of the other sidecar migrations, but nothing that I am aware has been tied to a specific Longhorn user issue. My main reason for trying to upgrade these sidecars in the 1.3.x and 1.4.x Longhorn release branches is that most of the versions we are using are no longer "supported" by the maintainers.

Included in the backported list is one major version upgrade (csi-attacher 3.4.0 -> 4.2.0). I have confirmed that the breaking change associated with that major version bump is transparent to Longhorn. Otherwise, we are looking at minor version upgrades.

To me, this is mostly a matter of policy. If the team generally prefers to be a bit more conservative about this type of thing, I can certainly respect that. I'll defer to you, @innobead, and/or other maintainers with more Longhorn experience. Let me know and I will close the 1.3.x and 1.4.x PRs.

**Comment by PhanLe1010** [comment]
Thanks @ejweber and @innobead for the conversation about updating `csi-xxx` components in 1.3.x and 1.4.x releases.

After reading through the comments, I can see the benefit of updating `csi-xxx` components in 1.3.x and 1.4.x releases which are to have minor bug fixes and move into the supported upstream versions.

However, I do see that there are important issues if we decide to do the backport:
1. The 1.3.x and 1.4.x have been marked as stable releases. Bumping csi-xxx components to their newer major versions will introduce risk and destabilize them. In the past, we had hit an accident in which bumping csi-provisioner version broke the volume's filesystem ownership and permission, and we had to release an emergency release for that accident. This risk is too high to have it into a Longhorn patch release of a stable release.
2. Even though we can test the PR using our e2e tests, it cannot cover all corner cases

Therefore, I agree that we should only bump the csi-xxx components in 1.5 release

**Comment by innobead** [comment]
> My main reason for trying to upgrade these sidecars in the 1.3.x and 1.4.x Longhorn release branches is that most of the versions we are using are no longer "supported" by the maintainers.

@ejweber Is there any support lifecycle for CSI sidecars? how did you know they are no longer supported?

Yes, as I mentioned and what @PhanLe1010 said, we only upgrade to patch versions of any external dependencies for existing release branches given assuming those components follow semantic versioning. So, let's close 1.3 & 1.4 backports but rather focus on 1.5/master only.

**Comment by PhanLe1010** [comment]
> how did you know they are no longer supported?

I think It is mentioned in their doc: 
https://kubernetes-csi.github.io/docs/external-provisioner.html#unsupported-versions

**Comment by ejweber** [comment]
Thanks, @PhanLe1010, that matrix (and the similar ones for other sidecars) was what I was referring to. The maintainers also describe their versioning and support policy here: https://kubernetes-csi.github.io/docs/project-policies.html#support. (Though IMO there are discrepancies between that policy and the matrices.)

**Comment by yangchiu** [comment]
Waiting for https://github.com/longhorn/longhorn-tests/pull/1357 and https://github.com/longhorn/longhorn-tests/pull/1371 being merged to fix failed test cases on master branch.

**Comment by yangchiu** [comment]
e2e testing results:

amd64: https://ci.longhorn.io/job/public/job/master/job/sles/job/amd64/job/longhorn-tests-sles-amd64/508/
amd64 upgrade: https://ci.longhorn.io/job/public/job/master/job/sles/job/amd64/job/longhorn-upgrade-tests-sles-amd64/487/
arm64: https://ci.longhorn.io/job/public/job/master/job/sles/job/arm64/job/longhorn-tests-sles-arm64/495/
arm64 upgrade: https://ci.longhorn.io/job/public/job/master/job/sles/job/arm64/job/longhorn-upgrade-tests-sles-arm64/494/

Basically there's no csi-related failed test cases except `test_csi_mount_volume_online_expansion` which is a newly added test case in https://github.com/longhorn/longhorn/issues/5839 and it's always failed on both `master` and `v1.4.x` since it's implemented. This failure probably isn't related to this ticket. @chriscchien Can you help to check the test case? Many thanks!

**Comment by yangchiu** [comment]
Close this and track failed test case `test_csi_mount_volume_online_expansion` on https://github.com/longhorn/longhorn/issues/5839
